src/include/common/concurrent_pointer_vector.h

```diff
@@ -0,0 +1,280 @@
+#pragma once
+
+#include <common/constants.h>
+#include <common/macros.h>
+#include <common/shared_latch.h>
+#include <execution/util/execution_common.h>
+#include <tbb/mutex.h>
+#include <atomic>              // NOLINT
+#include <condition_variable>  // NOLINT
+
+namespace terrier::common {
+
+/**
+ * A concurrent Vector of pointers that supports inserts and accesses at an index
+ */
+template <class T>
+class alignas(Constants::CACHELINE_SIZE) ConcurrentPointerVector {
+ public:
+  /**
+   * ConcurrentPointerVector constuctor
+   *
+   * @param start_size optional initial allocation size for the vector. Default is START_SIZE
+   */
+  explicit ConcurrentPointerVector(uint64_t start_size = START_SIZE)
+      : capacity_(start_size == 0 ? 1 : start_size), claimable_index_(0), first_not_readable_index_(0) {
+    array_ = new T *[capacity_];
+    for (uint64_t i = 0; i < capacity_; i++) SetNotReadable(array_, i);
+  }
+
+  /**
+   * ~ConcurrentPointerVector deconstructor
+   */
+  ~ConcurrentPointerVector() { delete[] array_; }
+
+  /**
+   * Insert
+   *
+   * Each new item atomically claims an index into which it will insert.
+   * Inserts handle resizes, and updating readability flags.
+   *
+   * @param item pointer to item to be inserted
+   * @return index of item in vector
+   */
+  uint64_t Insert(T *item) {
+    // claims index in vector
+    uint64_t my_index;
+    do {
+      my_index = claimable_index_;
+    } while (!claimable_index_.compare_exchange_strong(my_index, my_index + 1));
+
+    // resize vector
+    while (my_index >= capacity_) {
+      if (UNLIKELY(my_index == capacity_)) {
```

Comment:
Could you add a comment on why this is safe? (Something about the compare xchg strong above perhaps)

Reply:
i added one let me know if it makes things more clear